src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -448,50 +437,38 @@ public static Type resolveBound(TypeVariable<?> typeVariable) {
     return bound == Object.class ? Unknown.class : bound;
   }
 
-  /**
-   * Resolves method ref offset.
-   */
-  private static int resolveMethodRefOffset(ConstantPool constantPool) {
-    int offset = METHOD_REF_OFFSET;
-
-    if (offset == -1) {
-      int constantPoolSize = constantPool.getSize();
-
-      for (int i = 0; i < constantPoolSize; i++) {
-        try {
-          constantPool.getMemberRefInfoAt(constantPoolSize - i);
-          offset = i;
-          break;
-        } catch (IllegalArgumentException ignore) {
-        }
-      }
-
-      METHOD_REF_OFFSET = offset;
-    }
-
-    if (offset >= 0)
-      return offset;
-    else
-      throw new MethodRefOffsetResolutionFailed();
+  private static final class MethodRefOffsetResolutionFailed extends Exception {
   }
 
   /**
-   * Resolves method ref offset of method reference lambda type having autoboxed return type. We can't cache the result
-   * value because results are different depending on given {@code lambdaType}.
+   * Resolves method ref info.
    */
-  private static int resolveAutoboxedMethodRefOffset(ConstantPool constantPool, Class<?> lambdaType) {
+  private static String[] getMethodRefInfo(ConstantPool constantPool) throws MethodRefOffsetResolutionFailed {
     int constantPoolSize = constantPool.getSize();
+    String[] returnValue = null;
 
-    for (int i = resolveMethodRefOffset(constantPool) + 1; i < constantPoolSize; i++) {
+    for (int i = 0; i < constantPoolSize; i++) {
       try {
-        // ignore constructor
-        String[] refs = constantPool.getMemberRefInfoAt(constantPoolSize - i);
-        if (!refs[1].equals("<init>"))
-          return i;
+        String[] methodRefInfo = constantPool.getMemberRefInfoAt(constantPoolSize - i);
+        String methodName = methodRefInfo[1];
+        // Always ignore constructors
+        if (methodName.equals("<init>")) {
+          continue;
+        }
+        // If we found a method named valueOf, keep searching for other methods but still return
+        // the information for the valueOf method in case it turns out to be the only valid method.
+        returnValue = methodRefInfo;
```

Comment:
If you always unconditionally skip `valueOf`, even if there's no better alternative, then this test seems to fail: https://github.com/jhalterman/typetools/blob/b264eb1473989f4967698e9ce95567e209fbee9d/src/test/java/net/jodah/typetools/functional/LambdaTest.java#L216